### PR TITLE
py-zfit: add v0.18.*

### PR DESCRIPTION
**Pull request by jonas-eschle**
Adding 0.18 that upgrades the TF version and relaxes the caching reset https://github.com/zfit/zfit/releases

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-zfit/package.py
Comment (vvolkl) [inline var/spack/repos/builtin/packages/py-zfit/package.py]: Shouldn't this be:
```suggestion
        depends_on("py-attrs", when="@0.15:")
```
?

Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py]: Exactly, that was a typo, thanks!

**Comment by alecbcs** [comment]
@jonas-eschle can you add a short description to the PR or link to the upstream release notes?

**Comment by vvolkl** [comment]
If you add a 
```
    tags = ["hep"]
```
line to the recipe, I'm happy to merge this PR.

File: var/spack/repos/builtin/packages/py-zfit/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/py-zfit/package.py:1]: ```depends_on("python@3.9:3.11", when="@0.18:")```

Comment (haampie) [inline var/spack/repos/builtin/packages/py-zfit/package.py:1]: uncomment
```
depends_on("py-tensorflow@2.15", when="@0.18:")
depends_on("py-tensorflow-probability@0.23", when="@0.18:")
```

**Comment by vvolkl** [comment]
I had some trouble building the dependencies, should be fixed by: https://github.com/spack/spack/pull/43246

File: var/spack/repos/builtin/packages/py-zfit/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: Is there something wrong with this line?

in case you would want to relax the requirement, I don't think it there is any gain to it: TF keeps a quite strict python version (and for good reasons, a new version is usually non-trivial for TF) and zfit keeps a rather strict TF version (for good reasons again, as things just break in between, we're using quite heavily the lower mechanics of TF)

**Comment by jonas-eschle** [comment]
> If you add a
> 
> ```
>     tags = ["hep"]
> ```

Ah, the tags seems to be new, nice! (I didn't really find docs on them though? Just a few github discussions?)  I've added a few that fit well (it's actually not a HEP specific library, although widely used in HEP), I didn't really find any existing tags on any of the subjects.

**Comment by jonas-eschle** [comment]
We'll just need to wait for https://github.com/spack/spack/pull/43249 to be in

File: var/spack/repos/builtin/packages/py-zfit/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: The upperbound is missing:

```
$ curl -LfsS 'https://pypi.org/pypi/zfit/0.18.2/json' | jq -r .info.requires_python
<3.12,>=3.9
```

Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: Maybe I am misunderstanding the spack syntax, but isn't `python@3.9:3.11` restricting it to be within 3.9 and 3.11 (inclusive?)

Could you make a suggestion on what you think should be written instead? Apologies if I am missing the point

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: @jonas-eschle Yes it looks like some misunderstanding: @0.18: no longer supports 3.8.

This is not about tensorflow, but about `zfit`'s expressed Python version requirement that changes from 0.17.x to 0.18.x to no longer support Python 3.8:

hampie showed:
```py
$ curl -LfsS 'https://pypi.org/pypi/zfit/0.18.2/json' | jq -r .info.requires_python
<3.12,>=3.9
```
Thus, after the existing line
```py
       depends_on("python@3.8:3.11", type=("build", "run"))
```
add this line:
```py
       depends_on("python@3.9:3.11", when="@0.18:")
```
To add this to the diff:
```py
        depends_on("python@3.8:3.11", type=("build", "run"))
+       depends_on("python@3.9:3.11", when="@0.18:")
```
This tells that `3.8` to `3.11` is fine for older version, but `zfit >= 0.18` needs `3.9` to `3.11`. The `type=("build", "run")` applies to all versions, thus `when="@0.18:"` only needs to add "python@3.9:3.11" to reject Python 3.8.x for 0.18:

This should make it clearly expressed and is a clear instruction now.

Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: Ah, I didn't realize that it was still 3.8 in the requirements, I misread that. All the rest is clear and we agree. Changed from 3.8 to 3.9

**Comment by bernhardkaindl** [comment]
> We'll just need to wait for #43249 to be in

@jonas-eschle Merged #43249 after 5 days (it is the time that we should wait for package maintainer comments)

File: var/spack/repos/builtin/packages/py-zfit/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-zfit/package.py:1]: @haampie and apologies for the many mistakes in this PR: I believed we had 0.18 already in and we're just patching the 0.18.2, but I confused that with the conda-forge merge and went too fast about things. Thanks for all the comments!

**Comment by jonas-eschle** [comment]
thanks a lot @bernhardkaindl , looks ready from my side, as soon as the CI passes, this can be merged

**Comment by jonas-eschle** [comment]
@bernhardkaindl this is ready from my side, can this be merged?

**Review by bernhardkaindl — APPROVED** [review]
This has nothing to do with your PR, but I get this error when I try to build this branch:
```py
==> Installing py-tensorboard-2.16.2-qbittes34bjqau3qk57lwdaxyvho7wvp [1/1]
==> Cannot find version 2.16.2 in url_list
```
But this PR has the same error: https://github.com/spack/spack/pull/43056
https://gitlab.spack.io/spack/spack/-/jobs/10497564

So as I see nothing wrong specifically with your update, even if I'd have liked to build it, I approve.

**Comment by jonas-eschle** [comment]
Yes, I experience the same error, but as you said, unrelated to this.

Can be merged from my side (I don't have the rights), but glad to wait until it can be built as well

**Review by bernhardkaindl — APPROVED** [review]
JFYI - This looks a lot better now:

@scheibelp | @tldahlgren: As discussed in the weekly spack call:

I rebased the branch of this PR for a local build to the current develop branch:

That fixes the error I and @jonas-eschle saw:
```
==> Cannot find version 2.15.2 in url_list
```
... when attempting to fetch `py-tensorboard@2.15.2`.
This was triggered when this PR is based on develop branch from March 23rd
(the branch of this PR is still based on it, but on merge, it will be on top of the latest develop)

@jonas-eschle: The build is now continuing to build `py-tensorflow@2.15.1`, and once the full build has succeeded, I'd be ready to merge too.

**Review by bernhardkaindl — APPROVED** [review]
Build of py-zfit-0.18.2 with py-tensorflow-2.15.1 and py-tensorboard@2.15.2 successful.

**Pull request merged**
